### ISSUE TITLE
关于数据划分的问题

### ISSUE THREAD

**Issue by HuYou523** [snailpt/CTNet#7]
请问老师，您的数据集划分是在dataloader之后，开始提取dataloader里面的数据时再进行训练集和验证集划分，这会不会导致每个epoch中，训练集和验证集划分随机性太大，因为dataloader里面的shuffle不是会打乱顺序吗？会不会出现在验证集中有一种标签的数据未被划分到的情况？感觉随机性太大，麻烦您有空解答一下，感谢

**Comment by snailpt** [comment]
这是个很好的问题，谢谢您的关注!
1. 我也发现这个问题了，CTNet项目里训练集和测试集的确实存在训练集和验证集划分太过随机的问题，从而导致性能不太稳定。更稳定的做法应该是划分验证集的时候，每个类别都要取出相应数量的训练样本做验证集。
2. CTNet在训练的时候做了数据增强，在那里确保了每个batch增强的数据里都有各个类别的合成训练数据，因此不会存在训练集里没有对应类别的情况。
如果项目对您有帮助，请给点个星，支持下，谢谢！

**Comment by HuYou523** [comment]
感谢您的回复！对于数据增强部分我有个疑惑，你是将训练集的所有数据进行数据增强然后拼接到划分后的训练集中，每一个batchsize都会进行一次增强，把整个训练集增强后拼接到这个batchsize上，感觉这样计算量大且这样会造成验证集提前泄露吗？

**Comment by snailpt** [comment]
每个batch里重新生成增强数据，会增加训练样本的多样性。
验证集确实是会有部分trial的部分segmentation会发生提前泄露，目前还没想到更好的方案，毕竟每个受试者的总的训练样量也就200-300个左右。欢迎加入QQ讨论组837800443。

**Comment by snailpt** [comment]
如果有更好的验证集的划分方式，也希望您可以提出来一起研究，共同进步！

**Comment by HuYou523** [comment]
感谢您的回复！我的想法是使用train_test_split来划分训练集和验证集，我今天试了一下，我在没有使用数据增强的情况下，准确率只有75%，性能下降了3%，个别个体损失值出现轻微上升趋势，我也不知道是否是我操作有误，想听听你的见解

**Comment by snailpt** [comment]
不做数据增强，模型没有得到充分的训练使得准确率下降。这是很正常的现象。性能下降3%啥意思？做了数据增强只有78%？

**Comment by HuYou523** [comment]
感谢回复！您原代码的结果我是复现的出来的，因为我实验的需要我把数据增强部分去除后进行了实验，准确率在77到79之间，然后我发现训练集和验证集划分可能会存在一点瑕疵，我就使用train_test_split来划分，准确率就在未数据增强的基础上下降2到3个点。我想问一下，原代码验证集损失低于训练集损失是正常的吗？我使用train_test_split来划分后，验证集损失是一直高于训练集损失的

**Comment by snailpt** [comment]
用train_test_split来划分，验证集里的trial里的各segmentation片段从未泄露，验证集损失一直高于训练集损失，这个正常现象吧. 原代码里有一个很重要的训练参数，就是数据增强的次数，我试过设置为3比设置为1会高不少

**Comment by HuYou523** [comment]
原代码验证集损失低于训练集损失，侧面说明了其实验证集是存在一点泄露吧？这样训练会不会不好进行改进模型呢？就是不知道为什么用了train_test_split后效果会掉2个点左右

**Comment by snailpt** [comment]
原代码的验证集写法上有小问题的，存在较大的泄露，没想到误打误撞居然效果还不错

---- 回复的原邮件 ----
| 发件人 | ***@***.***> |
| 发送日期 | 2025年04月19日 09:35 |
| 收件人 | snailpt/CTNet ***@***.***> |
| 抄送人 | snailpt ***@***.***>,
Comment ***@***.***> |
| 主题 | Re: [snailpt/CTNet] 关于数据划分的问题 (Issue #7) |

原代码验证集损失低于训练集损失，侧面说明了其实验证集是存在一点泄露吧？这样训练会不会不好进行改进模型呢？就是不知道为什么用了train_test_split后效果会掉2个点左右

—
Reply to this email directly, view it on GitHub, or unsubscribe.
You are receiving this because you commented.Message ID: ***@***.***>

HuYou523 left a comment (snailpt/CTNet#7)

原代码验证集损失低于训练集损失，侧面说明了其实验证集是存在一点泄露吧？这样训练会不会不好进行改进模型呢？就是不知道为什么用了train_test_split后效果会掉2个点左右

**Comment by HuYou523** [comment]
非常奇妙的现象，完全不知道为什么

**Comment by snailpt** [comment]
我原本训练和验证集的想法是7:3开，但是写代码的时候没发现问题：“val_label_list.append(label[number_validate:]”应该是val_label_list.append(label[-number_validate:]才对。说明我原先的代码里的验证集估计有一半是提前泄露过的。至于改过来小面积泄露效果如何，我还没运行。
从写论文的角度，其实这个问题不用纠结，选定一个训练集和验证集的标准，然后所有的模型用同一套标准来对比，这样还是很公平的。至于CTNet用了新标准准确率下降了，也没关系，因为其他模型可能也同样会下降的。然后写论文引用CTNet的时候，您具体算出来多少就写多少，没关系的。
感谢您对项目的关注，非常期待您的其他建议及意见。

**Comment by HuYou523** [comment]
懂了！感谢您的耐心回复解决我的疑惑！受益匪浅

**Comment by HuYou523** [comment]
你说的这个问题我刚开始看代码的时候已经发现并改进了，其实貌似没多大区别
